### ISSUE TITLE
[3.3.5][6.x] Implement real time statistic visualization in TrinityCore

### ISSUE THREAD

**Issue by DDuarte** [TrinityCore/TrinityCore#15075]
### Software:
- https://influxdb.com/ Time-series oriented database
- http://grafana.org/ Dashboard/graphs visualization
- https://github.com/DDuarte/influxdb-cpp C++ InfluxDB driver (TODO)

Alternatives to InfluxDB: http://graphite.wikidot.com/ (also supported by Grafana)

#### Technical oriented
- I/O networking traffic:
    - Packets sent
    - Packets received
    - Average ping
    - Traffic in
    - Traffic out
- World session update time
- Map update time
- Database async queries queued count
- Server uptime
- Active connections
- Queued connections
- CPU usage
- RAM usage

#### Game oriented
- Players online (per faction)
- Logins per hour, per day, day of week, etc
- Mails sent
- Auction house usage
- Character levels
- Heatmap of players position
- Gold earn/spend
- LFG queues

### Notes / implementation details:
- Batch inserts to the DB (i.e keep X (10/100?) events in memory and then send them at once)
- This feature is absolutely optional and disabled by default
- Tracking code should have low impact in performance
- It should be possible to enable and disable statistic code at runtime (worldserver.conf + reloads)
- The data could be written at each world update loop or using asio at configurable intervals

**Comment by jackpoz** [comment]
I'm saving some notes about this issue at https://github.com/jackpoz/TrinityCore/issues/41 and developing it at https://github.com/jackpoz/TrinityCore/tree/3.3.5-Graphana .
In addition to "which statistics should be saved ?" we should define "which graphs/dashboard should we setup ?"

**Comment by jackpoz** [comment]
> Logins per hour, per day, day of week, etc

this is achieved by saving the login event and then creating a dashboard with it, specifying "sum" function in the grafana graph

> Heatmap of players position

I'm not sure a heatmap fits in time-based series, especially if that map should have the look of WoW map

**Comment by jackpoz** [comment]
Few useful steps to setup InfluxDB & Grafana:
- Create the "worldserver" database from InfluxDB CLI:
 - CREATE DATABASE worldserver
 - CREATE USER grafana WITH PASSWORD 'grafana'
 - GRANT READ ON worldserver TO grafana

- Add the database to Grafana
![image](https://cloud.githubusercontent.com/assets/1153754/13438973/b2699b36-dfeb-11e5-8858-c747293e7a8b.png)

- Import the dashboard
 - [Windows realm.txt](https://github.com/TrinityCore/TrinityCore/files/153458/Windows.realm.txt)
 - Fix the datasource of each chart

**Comment by jackpoz** [comment]
According to https://github.com/influxdata/influxdb/issues/5880 latest InfluxDB version doesn't support "Connection: keep-alive" anymore and always closes the connection.

**Comment by jackpoz** [comment]
>we can always add a special api for performance critical code so a local queue gets created and it's passed to the stats logger later. so instead of queueing 1 event, we'll queue a batch of events

**Comment by jackpoz** [comment]
A counter class could be used to automatically call TC_STATS_VALUE() with specified arguments either on a fixed point or when going out of scope or both (with flag to send it only once). This will keep code cleaner from random uint32 counters.

**Comment by jackpoz** [comment]
Database related metrics could be monitored using Telegraf https://influxdata.com/blog/announcing-telegraf-a-metrics-collector-for-influxdb/

>Telegraf can collect metrics for system (CPU, memory, network, etc), Docker, Redis, MySQL, and PostgreSQL.

**Comment by jackpoz** [comment]
- [ ] rename Stats to Metrics
- [ ] remove every reference to InfluxDB from configs
- [ ] add graphs to /contrib

**Comment by jackpoz** [comment]
http://grafana.org/community/percona_webinar.html "Using Grafana for MySQL Monitoring"

**Comment by Nawuko** [comment]
They talking about Monitoring an MySQL-Server not MySQL as datasource? Or are they?
